### ISSUE TITLE
test_maxwell_tensor_contravariant_for_natural_units fails on Debian arm64

### ISSUE THREAD

**Issue by olebole** [einsteinpy/einsteinpy#546]
🐞 **Problem**

The Debian Einsteinpy package fails a test on the arm64 platform on Debian:

```
_____________ test_maxwell_tensor_contravariant_for_natural_units ______________

    def test_maxwell_tensor_contravariant_for_natural_units():
        #  Theoretical background is required to write extensive tests. Right now only skew-symmetric property is being checked.
        M = 1e22
        r = 5.5
        theta = 2 * np.pi / 5
        Q = 45.0
        a = 0.7
        m = kerrnewman_utils.maxwell_tensor_contravariant(r, theta, a, Q, M, 1.0, 1.0, 1.0)
        for t in range(16):
            i = int(t / 4) % 4
            j = t % 4
>           assert_allclose(0, m[i, j] + m[j, i], rtol=0.0, atol=1e-10)
E           AssertionError: 
E           Not equal to tolerance rtol=0, atol=1e-10
E           
E           Mismatched elements: 1 / 1 (100%)
E           Max absolute difference: 1.
E           Max relative difference: 1.
E            x: array(0)
E            y: array(-1.)

/usr/lib/python3/dist-packages/einsteinpy/tests/test_utils/test_kerrnewman_utils.py:145: AssertionError
``` 
Full test log is [here](https://ci.debian.net/data/autopkgtest/testing/arm64/e/einsteinpy/8149630/log.gz). The same failure [happens on the ppc64el platform](https://ci.debian.net/data/autopkgtest/testing/ppc64el/e/einsteinpy/8150246/log.gz).
This failure prevents the package from migrating to Debian Testing, which is required to be included in the next Debian release.

Python package versions (from the log file):

Package | Version
------------|---------
astropy | 4.1-2
atomicwrites | 1.4.0-1
attr | 20.1.0-1
certifi | 2020.6.20-1
chardet | 3.0.4-7
configobj | 5.0.6-4
cov-core | 1.15.0-3
coverage | 5.1+dfsg.1-2+b1
cycler | 0.10.0-3
dateutil | 2.8.1-4
decorator | 4.4.2-2
distutils | 3.8.6-1
hypothesis | 5.41.2-1
idna | 2.10-1
importlib-metadata | 1.6.0-2
ipython-genutils | 0.2.0-4
jsonschema | 3.2.0-3
jupyter-core | 4.6.3-4
kiwisolver | 1.2.0-1+b1
lib2to3 | 3.8.6-1
matplotlib | 3.3.2-2
more-itertools | 4.2.0-3
mpmath | 1.1.0-2
nbformat | 5.0.8-1
nose2 | 0.9.2-1
numpy | 1:1.19.4-1
packaging | 20.4-1
pil | 8.0.1-1
pip | 20.1.1-2
pkg-resources | 50.3.0-1
plotly | 4.12.0+dfsg-1
pluggy | 0.13.0-5
ply | 3.11-4
psutil | 5.7.3-1
py | 1.9.0-1
pyparsing | 2.4.7-1
pyrsistent | 0.15.5-1+b2
pytest | 4.6.11-1
pytest-arraydiff | 0.3-1
pytest-astropy | 0.8.0-1
pytest-astropy-header | 0.1.2-3
pytest-cov | 2.8.1-1
pytest-doctestplus | 0.7.0-1
pytest-filter-subpackage | 0.1.1-2
pytest-openfiles | 0.5.0-1
pytest-remotedata | 0.3.2-1
requests | 2.24.0+dfsg-1
retrying | 1.3.3-4
scipy | 1.5.3-2
setuptools | 50.3.0-1
six | 1.15.0-1
sortedcontainers | 2.1.0-2
sympy | 1.6.2-2
traitlets | 5.0.5-1
tz | 2020.1-2
urllib3 | 1.25.9-1
wcwidth | 0.1.9+dfsg1-2
wheel | 0.34.2-1
zipp | 1.0.0-3
python |  3.8.6-1

@shreyasbapat If this is fixed, it would be nice to have a new Debian package that I could upload.

**Comment by godslayer201** [comment]
Hey @olebole I'll try to replicate the issue on my local machine and try to get this fixed. 
 

We think it's a architecture issue.

**Comment by JeS24** [comment]
A couple of observations:

1. All tests pass with `einsteinpy-0.3.0` (no `numba`) on a Raspberry Pi 4 (`armhf` - 32-bit), running Raspbian 10 (fresh installation | up-to-date). Same result using PyDroid 3 (all EPy dependencies up-to-date) on a Android 10 phone (`arm64`) . This prima facie rules out architecture issues.

2. The `m` array in that test contains large and small values, as can be seen below:
```
array([[ 0.00000000e+00,  8.22147733e+15,  3.91995917e-09,  0.00000000e+00],
       [-8.22147733e+15,  0.00000000e+00,  0.00000000e+00, -1.29849168e+16],
       [-3.91995917e-09,  0.00000000e+00,  0.00000000e+00, -6.19114324e-09],
       [ 0.00000000e+00,  1.29849168e+16,  6.19114324e-09,  0.00000000e+00]])
```
At this point, I am not sure about the origin of this error.

**Comment by olebole** [comment]
We observe the following results:

 * amd64: Pass
 * arm64: Fail
 * armhf: Pass
 * i386: Pass
 * ppc64el: Fail
 * s390x: Fail

The arm64 test [fails already since november](https://ci.debian.net/packages/e/einsteinpy/unstable/arm64/); the two others were tested only once (~one week ago).

> At this point, I am not sure about the origin of this error.

What I could imagine: the assert subtracts two almost-equat numbers, which may be (as shown in your example) quite large:

    assert_allclose(0, m[i, j] + m[j, i], rtol=0.0, atol=1e-10)

specifically, there are values with 15-16 digits. This is on the edge of accuracy for doubles, and tiny deviations in the calculation may lead to a change in the lowest digit, which is in the order of one. IMO this is the reason here.
I would rather check that the array is antisymmetric, by replacing [the whole `for` loop](https://github.com/einsteinpy/einsteinpy/blob/8cd0d7f3a372d462d5bf0217f62234e9cf8839e9/src/einsteinpy/tests/test_utils/test_kerrnewman_utils.py#L142-L145) with

    assert_allclose(m, -m.T, rtol=1e-12)

or so. This should be also faster, as it avoids the loop. Similarly probably here: https://github.com/einsteinpy/einsteinpy/blob/8cd0d7f3a372d462d5bf0217f62234e9cf8839e9/src/einsteinpy/tests/test_utils/test_kerrnewman_utils.py#L113-L116

(and maybe on other places). Generally, I would recommend to avoid subtracting large numbers; this is always a cause of pain.

**Comment by shreyasbapat** [comment]
That's correct. How do we make changes to it @olebole ?

**Comment by olebole** [comment]
@shreyasbapat I am not sure I understand what you mean…
I thought, one could just fix this (f.e. with a Pull request), and then create a new version, for which you could update the Debian package, which I then would upload.

**Comment by shreyasbapat** [comment]
What I meant was, since there is no actual way for us to see if our pull request actually fixes the issue, how exactly do we test it?

**Comment by olebole** [comment]
The easiest way is to create a Debian package with the fix (which I just did - 0.3.0-2). Let's see whether this fixes it.

**Comment by olebole** [comment]
[This patch](https://salsa.debian.org/debian-astro-team/einsteinpy/-/blob/master/debian/patches/Replace-subtraction-of-large-values-in-test-with-direct-c.patch) fixes the CI test on the [arm64](https://ci.debian.net/data/autopkgtest/unstable/arm64/e/einsteinpy/9525573/log.gz) and [ppc64el](https://ci.debian.net/data/autopkgtest/unstable/ppc64el/e/einsteinpy/9518030/log.gz) platforms, so I would think that this is the right fix of the issue.
@shreyasbapat if you would create a new version for Debian, I would happily upload it.

**Comment by shreyasbapat** [comment]
Alright, I'll fix this and create a release 0.3.1 on pypi :D This Friday

**Comment by olebole** [comment]
Can you also update the Debian package?

**Comment by shreyasbapat** [comment]
> Can you also update the Debian package? (On Salsa, I would upload it then)

Sure. I don't remember how to do that exactly but I will do it

**Comment by shreyasbapat** [comment]
I am preparing a patch and a patch release now

**Comment by shreyasbapat** [comment]
So, interestingly, we have already removed this test from the current `main`. And this will not be present in the upcoming `0.4.0` release. So, I just need to patch it for `0.3.1`

**Comment by shreyasbapat** [comment]
I have released EinsteinPy 0.3.1 : https://pypi.org/project/einsteinpy/

**Comment by shreyasbapat** [comment]
I will keep posting here about the Debian Salsa Updates
